Automattic/wp-calypso
# State: Add state behaviors for setting site icon

Blocked by ~#9675~
~Cherry-picks 215ce3e036de1bd78969e3b3f1f50c7ffef223df from #9706~
~Partially cherry-picks f61daf539db322061ea77d0e139e783e4adc8f66 from #9706~

This pull request seeks to trigger the network request and state updates to update a site's icon after having selected and cropped an image from the media library. While the included changes will permanently update the site option for site icon, because the option is not yet respected yet on the server-side, the effect is only temporary (see Caveats).

__Testing instructions:__

1. Navigate to [site settings](http://calypso.localhost:3000/settings)
2. Select a site, if prompted
3. Click Change under Site Icon
4. Select a media item
5. Click Continue
6. Resize image
7. Click Done
8. Note that after upload completes, the site icon on settings screen updates to reflect site icon

Extra credit: Confirm from [developer tools console](https://developer.wordpress.com/docs/api/console/) that upon requesting `/sites/%s/settings` (substituting your site) that the `settings.site_icon` value reflects the ID of the newly uploaded media.

__Caveats:__

Caveats are aplenty and many of which are planned to be addressed in parallel pull requests:

- The sidebar `<CurrentSite />` component will not update its site icon because it is still bound to the `sites-list` store. We may need to update this component to use Redux state and/or add a compatibility layer between updating site settings in Redux to `sites-list`.
- The `<SiteIcon />` in the site settings panel does not update immediately upon clicking Done because there is an issue in the component where it tries to append a query argument to the local file URL
- Refreshing the page will revert the site icon to its original value (there is still work to be done on the server-side to respect the site option)
- There is no notice shown upon successful upload (blocked by #9085, cc @youknowriad)
